src/Build.UnitTests/Evaluation/Expander_Tests.cs

```diff
@@ -2850,6 +2851,18 @@ public void PropertyFunctionRuntimeInformation(string propertyFunction, string e
             Assert.Equal(expectedExpansion, result);
         }
 
+        [Fact]
+        public void TestFeatureOSAPIs()
```

Comment:
The initial feature request #5982 was to add some new property functions where they're available, so users can use the new functions in their MSBuild logic or conditions.

You've done that, by adding `OperatingSystem` to the "available types". So there should be some new expressions, like `$([System.OperatingSystem]::IsWindows())`, that should work now. I'd like to see some tests that use those expressions.

This helps with two things:

1. It guards against regressions, accidentally removing those APIs.
2. It shows us in this PR what the user experience for the expressions is.

Reply:
It looks like that's failing on the `net472` binaries of MSBuild, which makes sense, since the API is new to .NET 5.

So now . . . what should we do about that? We haven't yet exposed any types' functions that are available only on one TargetFramework. The easiest thing would be to make the tests conditional as well (with `[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "OperatingSystem is .NET 5.0+")]`).

However, I'd like to improve the experience when they're used on .NET Framework MSBuild (like "in Visual Studio", a common use case :)). The most complicated option would be to reimplement all of the APIs so that the property function versions of them worked everywhere, but that might be a fair amount of code to maintain. Though maybe we could hardcode a bunch of it: `IsWindows=true` and everything else is false, for instance.

Or we could try to have a better error that says something like `OperatingSystem methods are not available on .NET Framework`.

@Forgind did you have a design for this in mind when you marked the bug as good-first-issue?

Reply:
I marked this as good-first-issue in 2020, so I have no recollection of what I was thinking. I suspect that I hadn't considered the net472 case and had assumed the PR would look just like what this one does now.

That said, I glanced through the list of APIs not available in net472, and only two of them looked like they can't be either "return false" or "return true", and those last two should be fairly easy to write, so I'm in favor of reimplementing them all.

Reply:
I agree, we can assume running on net472 means running on Windows (i.e. ignore the Mono msbuild use case).

For the reimplementation of `IsWindowsVersionAtLeast()` and we can just borrow the code from here: https://github.com/dotnet/runtime/blob/8c496e71e9617db81dc2c2747cf1d74bc1a84ad2/src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs#L305-L327

Reply:
Thanks for the assistance, this turned out to be a little more complicated of an error that I originally thought, but I would be more than happy to work through it with everyone involved and eventually get it merged and closed out and make sure I am understanding the change

Where would be the best place to reimplement? The project where these tests are associated with? 

Those inline tests were mimicked just as you described, IsWindows() = true, IsMac() = false, IsLinux() = false.

So since this is not available as a property function and you described an easier solution beyond making the InLineData tests that I was working on conditional. The target framework is currently .NET 5.0+ which currently allows backwards compatibility if implemented correctly for our .net4.7.2 error right here?

Reply:
> this turned out to be a little more complicated of an error that I originally thought

Yes, sorry about that!

> but I would be more than happy to work through it with everyone involved and eventually get it merged and closed out and make sure I am understanding the change

Fabulous! We're not in a big rush to have this feature and we can work with you along the way.

I think the easiest way to implement this would be to create a new type named `System.OperatingSystem` here in our codebase, with `#if` around it so it only compiles for .NET Framework.

We do this for a couple of other things, for example for `SupportedOSPlatform`:

https://github.com/dotnet/msbuild/blob/3777dcaf7edb3e86a070037ba53e742dd1872873/src/Framework/SupportedOSPlatform.cs#L4-L6

That should make all the code compile without the `#if`s in `Constants.cs`, but it won't make the tests pass until you've also implemented all of the methods of `OperatingSystem` in the simplified form we're thinking of where you can hardcode most of them to be the right return value on Windows.

Reply:
> For the reimplementation of `IsWindowsVersionAtLeast()` we can just borrow the code from here: https://github.com/dotnet/runtime/blob/8c496e71e9617db81dc2c2747cf1d74bc1a84ad2/src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs#L305-L327

Unfortunately I don't think we can, because it uses `Environment.OSVersion`, which differs on Windows between net472 and net5.0+: https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/5.0/environment-osversion-returns-correct-version

@WingZer0123 for now make that method throw a `PlatformNotSupportedException` while we figure out what to do.

Reply:
Am I correct in the checking the references for SupportOSPlatform.

These seems quite similar to what you are describing. 

`    [SupportedOSPlatform("windows")]
    private static bool IsLongPathsEnabledRegistry()
    {
        using (RegistryKey fileSystemKey = Registry.LocalMachine.OpenSubKey(WINDOWS_FILE_SYSTEM_REGISTRY_KEY))
        {
            object longPathsEnabledValue = fileSystemKey?.GetValue(WINDOWS_LONG_PATHS_ENABLED_VALUE_NAME, 0);
            return fileSystemKey != null && Convert.ToInt32(longPathsEnabledValue) == 1;
        }
    }

    /// <summary>
    /// Cached value for IsUnixLike (this method is called frequently during evaluation).
    /// </summary>
    private static readonly bool s_isUnixLike = IsLinux || IsOSX || IsBSD;

    /// <summary>
    /// Gets a flag indicating if we are running under a Unix-like system (Mac, Linux, etc.)
    /// </summary>
    internal static bool IsUnixLike
    {
        get { return s_isUnixLike; }
    }

    /// <summary>
    /// Gets a flag indicating if we are running under Linux
    /// </summary>
    [SupportedOSPlatformGuard("linux")]
    internal static bool IsLinux
    {
#if CLR2COMPATIBILITY
            get { return false; }
#else
        get { return RuntimeInformation.IsOSPlatform(OSPlatform.Linux); }
#endif
    }

    /// <summary>
    /// Gets a flag indicating if we are running under flavor of BSD (NetBSD, OpenBSD, FreeBSD)
    /// </summary>
    internal static bool IsBSD
    {
#if CLR2COMPATIBILITY
            get { return false; }
#else
        get
        {
            return RuntimeInformation.IsOSPlatform(OSPlatform.Create("FREEBSD")) ||
                   RuntimeInformation.IsOSPlatform(OSPlatform.Create("NETBSD")) ||
                   RuntimeInformation.IsOSPlatform(OSPlatform.Create("OPENBSD"));
        }
#endif
    }`

Reply:
@rainersigwald the `Environment.OSVersion` should work fine even for net472 since msbuild is opting in to the compatibility manifest here: https://github.com/dotnet/msbuild/blob/9bcc06cbe19ae2482ab18eab90a82fd079b26897/src/MSBuild/MSBuild.exe.manifest#L22-L23

Reply:
I noticed that doesn't include Windows 11, but I think Windows 11 was basically a shell over Windows 10, so that's fine, right?

Reply:
Yes that's fine, `8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a` means both Win10 and Win11 according to https://learn.microsoft.com/en-us/windows/win32/sysinfo/targeting-your-application-at-windows-8-1